File: src/Maximum subarray sum 2/Code/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    // Use long to avoid overflow of sums
    public long maxSubarrSum(int[] arr, int a, int b) {
        int n = arr.length;
        if (n == 0) return 0L;

        long[] prefix = new long[n + 1];
        prefix[0] = 0L;
        for (int i = 1; i <= n; ++i) prefix[i] = prefix[i - 1] + (long)arr[i - 1];

        Deque<Integer> dq = new ArrayDeque<>();
        long ans = Long.MIN_VALUE;

        for (int r = 1; r <= n; ++r) {
            int leftBound = Math.max(0, r - b);
            int newIdx = r - a;

            // remove indices that are out of the window
            while (!dq.isEmpty() && dq.peekFirst() < leftBound) dq.pollFirst();

            // add new index (if available) maintaining increasing prefix values
            if (newIdx >= 0) {
                while (!dq.isEmpty() && prefix[dq.peekLast()] >= prefix[newIdx]) dq.pollLast();
                dq.offerLast(newIdx);
            }

            if (!dq.isEmpty()) {
                ans = Math.max(ans, prefix[r] - prefix[dq.peekFirst()]);
            }
        }
        return ans;
    }
}

File: src/Maximum subarray sum 2/Code/solution.js
/**
 * @param {number[]} arr
 * @param {number} a
 * @param {number} b
 * @returns {number}
 */
class Solution {
  maxSubarrSum(arr, a, b) {
    const n = arr.length;
    if (n === 0) return 0;

    // prefix sums
    const prefix = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; ++i) prefix[i] = prefix[i - 1] + arr[i - 1];

    // deque implemented with array and head pointer for O(1) amortized front pop
    const dq = [];
    let head = 0;
    let ans = Number.NEGATIVE_INFINITY;

    for (let r = 1; r <= n; ++r) {
      const leftBound = Math.max(0, r - b);
      const newIdx = r - a;

      // pop front while out of window
      while (head < dq.length && dq[head] < leftBound) head++;

      // add new index (if valid) maintaining increasing prefix values
      if (newIdx >= 0) {
        while (
          head < dq.length &&
          prefix[dq[dq.length - 1]] >= prefix[newIdx]
        ) {
          dq.pop();
        }
        dq.push(newIdx);
      }

      if (head < dq.length) {
        const candidate = prefix[r] - prefix[dq[head]];
        if (candidate > ans) ans = candidate;
      }
    }

    return ans;
  }
}
